server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -164,16 +169,21 @@ private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSeg
          * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
          * IllegalStateException to fail the shard
          */
-        if (diff.different.isEmpty() == false) {
+
+        if (checkpointFromNewPrimary == false && diff.different.isEmpty() == false) {
             getFilesListener.onFailure(
                 new IllegalStateException(
                     new ParameterizedMessage("Shard {} has local copies of segments that differ from the primary", indexShard.shardId())
                         .getFormattedMessage()
                 )
             );
         }
-        final List<StoreFileMetadata> filesToFetch = new ArrayList<StoreFileMetadata>(diff.missing);
 
+        final List<StoreFileMetadata> filesToFetch = new ArrayList<StoreFileMetadata>(diff.missing);
+        if (checkpointFromNewPrimary && diff.different.isEmpty() == false) {
```

Comment:
Nit: Small comment/doc might be helpful to understand why is this needed.

Reply:
Updated.